### PR TITLE
[Flux] Add batched inference

### PR DESCRIPTION
**Pull request by CarlosGomes98**
This PR adds the ability to perform batched and multi-gpu inference on the Flux model, following from #1205 

- The main modification is to functions in `sampling.py`, allowing them to take several prompts and infer with them in a batch.
- Added a `infer.py` script which performs inference. This leverages the Trainer, which is great for reusability, but not so great because it inherits all the requirements of the trainer (such as a train dataset, which doesnt really make sense here)
- Added a `run_inference.sh` script and documented it in README.md
- Added an example file with prompts `prompts.txt`

### CODE REVIEW

**Comment by CarlosGomes98** [comment]
@wwwjn @tianyu-l Here is the split up inference bit :)

File: torchtitan/experiments/flux/sampling.py [hunk 1]
```diff
@@ -177,9 +202,9 @@ def denoise(
     # create positional encodings
     POSITION_DIM = 3
     latent_pos_enc = create_position_encoding_for_latents(
-        bsz, latent_height, latent_width, POSITION_DIM
+        1, latent_height, latent_width, POSITION_DIM
```
Comment (wwwjn) [inline torchtitan/experiments/flux/sampling.py:208]: QQ: Why we change the `bsz` to 1 here and later, as we are taking `bsz` prompts as input?

File: torchtitan/experiments/flux/sampling.py [hunk 2]
```diff
@@ -203,9 +231,12 @@ def denoise(
         if enable_classifer_free_guidance:
             pred_u, pred_c = pred.chunk(2)
             pred = pred_u + classifier_free_guidance_scale * (pred_c - pred_u)
-
+            pred = pred.repeat(2, 1, 1)
```
Comment (wwwjn) [inline torchtitan/experiments/flux/sampling.py:234]: And QQ: Why we need to repeat the first dimension of `pred`?

File: torchtitan/experiments/flux/scripts/infer.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            for rank_index in range(world_size):
+                global_idx = rank_index + in_rank_index * world_size
+                if global_idx >= total_prompts:
+                    break
+                all_images[global_idx] = gathered_images[rank_index][in_rank_index]
+        logger.info("Inference done")
+
+        pil_images = [torch_to_pil(img) for img in all_images]
+        if config.inference.save_path:
+            path = Path(config.job.dump_folder, config.inference.save_path)
```
Comment (wwwjn) [inline torchtitan/experiments/flux/scripts/infer.py:136]: Can we reuse the `save_image()` function in `sampling.py` here?

File: torchtitan/experiments/flux/scripts/infer.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+                clip_tokenizer=clip_tokenizer,
+                t5_encoder=trainer.t5_encoder,
+                clip_encoder=trainer.clip_encoder,
+            )
+            results.append(images.detach())
+    results = torch.cat(results, dim=0)
+    return results
+
+
+if __name__ == "__main__":
```
Comment (fegin) [inline torchtitan/experiments/flux/scripts/infer.py:61]: It's better to put the following logic in a separate function (like `main()`). This will allow easier logic reuse (e.g., for unittests).

File: torchtitan/experiments/flux/scripts/infer.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        )
+        # pad the outputs to make sure all ranks have the same number of images for the gather step
+        images = torch.cat(
+            [
+                images,
+                torch.zeros(
+                    math.ceil(total_prompts / world_size) - images.shape[0],
+                    3,
+                    256,
+                    256,
```
Comment (fegin) [inline torchtitan/experiments/flux/scripts/infer.py:100]: Can we make these magic numbers constant variables for the readability purpose?

File: torchtitan/experiments/flux/scripts/infer.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    return results
+
+
+if __name__ == "__main__":
+    init_logger()
+    config_manager = ConfigManager()
+    config = config_manager.parse_args()
+    trainer = FluxTrainer(config)
+    world_size = int(os.environ["WORLD_SIZE"])
+    global_id = int(os.environ["RANK"])
```
Comment (fegin) [inline torchtitan/experiments/flux/scripts/infer.py:67]: Can we use `rank` to match the convention?

File: torchtitan/experiments/flux/scripts/infer.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    prompts = original_prompts[global_id::world_size]
+
+    trainer.checkpointer.load(step=config.checkpoint.load_step)
+    t5_tokenizer = FluxTokenizer(
+        config.encoder.t5_encoder,
+        max_length=config.encoder.max_t5_encoding_len,
+    )
+    clip_tokenizer = FluxTokenizer(config.encoder.clip_encoder, max_length=77)
+
+    if global_id == 0:
```
Comment (fegin) [inline torchtitan/experiments/flux/scripts/infer.py:81]: This is not required, the logging information should be controlled by TorchRun configuration, which TorchTitan run scripts default to rank 0 only.

File: torchtitan/experiments/flux/scripts/infer.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+            math.ceil(total_prompts / world_size), 3, 256, 256, device=trainer.device
+        )
+
+    # Create a list of tensors to gather results
+    gathered_images = [
+        torch.zeros_like(images, device=trainer.device) for _ in range(world_size)
+    ]
+
+    # Gather images from all processes
+    torch.distributed.all_gather(gathered_images, images)
```
Comment (fegin) [inline torchtitan/experiments/flux/scripts/infer.py:117]: This should be `gather()` not `all_gather()` as you are not using the results on other ranks. Though I don't know if there will be any performance gains, `gather()` produces less total network traffic.

File: torchtitan/experiments/flux/scripts/infer.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+    # re-order the images to match the original ordering of prompts
+    if global_id == 0:
```
Comment (fegin) [inline torchtitan/experiments/flux/scripts/infer.py:120]: Another good motivation of making the logic in `main()` -- you can do early return here to remove one indention.

File: torchtitan/experiments/flux/scripts/infer.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            for i, img in enumerate(pil_images):
+                exif_data = Image.Exif()
+                exif_data[ExifTags.Base.Software] = "AI generated;txt2img;flux"
+                exif_data[ExifTags.Base.Make] = "Black Forest Labs"
+                exif_data[ExifTags.Base.Model] = "Schnell"
+                exif_data[ExifTags.Base.ImageDescription] = original_prompts[i]
+                img.save(
+                    path / f"img_{i}.png", exif=exif_data, quality=95, subsampling=0
+                )
+    torch.distributed.destroy_process_group()
```
Comment (fegin) [inline torchtitan/experiments/flux/scripts/infer.py:147]: We should something like:

```
try:
     main()
finally:
     if torch.distributed.is_initialized():
         torch.distributed.destroy_process_group()
```

**Review by fegin — COMMENTED** [review]
If `trainer` makes some assumptions about the dataset, we can rethink about the requirements of `trainer`. It's time to make `trainer` `inference/eval` friendly as well. cc., @wwwjn

File: torchtitan/experiments/flux/run_inference.sh [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+set -ex
+
+# use envs as local overrides for convenience
+# e.g.
+# LOG_RANK=0,1 NGPU=4 ./torchtitan/experiments/flux/run_inference.sh
+
+if [ -z "${JOB_FOLDER}" ]; then
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/run_inference.sh:15]: can we default to `job.dump_folder`'s default?

File: torchtitan/experiments/flux/scripts/infer.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+            path = Path(config.job.dump_folder, config.inference.save_path)
+            path.mkdir(parents=True, exist_ok=True)
+            for i, img in enumerate(pil_images):
+                exif_data = Image.Exif()
+                exif_data[ExifTags.Base.Software] = "AI generated;txt2img;flux"
+                exif_data[ExifTags.Base.Make] = "Black Forest Labs"
+                exif_data[ExifTags.Base.Model] = "Schnell"
+                exif_data[ExifTags.Base.ImageDescription] = original_prompts[i]
+                img.save(
+                    path / f"img_{i}.png", exif=exif_data, quality=95, subsampling=0
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/scripts/infer.py:145]: If eventually we are saving individual image files, why do we even perform gather / all-gather? We could save different images in the same folder from different ranks, just with unique names i.e. `rank_{i}` in the `.png` name.

File: torchtitan/experiments/flux/scripts/infer.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+
+def torch_to_pil(x: torch.Tensor) -> Image.Image:
+    x = x.clamp(-1, 1)
+    x = rearrange(x, "c h w -> h w c")
+    return Image.fromarray((127.5 * (x + 1.0)).cpu().byte().numpy())
+
+
+@record
+def inference(
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/scripts/infer.py:31]: not obvious to me why this is worth a standalone function -- can we just call `generate_image` in the main script?

File: torchtitan/experiments/flux/scripts/infer.py
Comment (tianyu-l) [inline torchtitan/experiments/flux/scripts/infer.py:1]: Functionality-wise, this seems similar to [`torchtitan/experiments/flux/tests/test_generate_image.py`](https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/tests/test_generate_image.py), but with parallelized model.
I think we can make `test_generate_image` a unit test, if not removing it after this multi-gpu generation lands. @wwwjn

File: torchtitan/experiments/flux/run_inference.sh
Comment (tianyu-l) [inline torchtitan/experiments/flux/run_inference.sh:1]: to group files a bit more logically, can we put `run_inference.sh`, `prompts.txt`, `infer.py` under the `flux/inference` folder? We can leave `sampling.py` outside as it's also used by the evaluation in `train.py`

File: torchtitan/experiments/flux/sampling.py [hunk 14]
```diff
@@ -203,9 +231,12 @@ def denoise(
...
         latents = latents + (t_prev - t_curr) * pred
 
+    if enable_classifer_free_guidance:
+        latents = latents.chunk(2)[1]
```
Comment (tianyu-l) [inline torchtitan/experiments/flux/sampling.py:238]: Yeah this code, together with the `pred = pred.repeat(2, 1, 1)` above, looks very obscure.
If they are necessary, please add adequate comments.

File: torchtitan/experiments/flux/sampling.py (hunk 1 above)
Comment (CarlosGomes98) [reply torchtitan/experiments/flux/sampling.py:208]: So theres 2 parts to this. The reason we can do this, is because for these particular tensors where im using 1 on the first dimension, they are the same for all samples. so we just want to repeat them for all of them. Due to torch broadcasting, whenever this is used in an operation with another tensor, this dimension will be expanded to match whatever is necessary from the other tensor (basically torch will automatically make this whatever the batch size is)

The reason we want to do this is twofold. 

1. It does save some memory to not have to carry around all these repeated tensors, but to just allow torch to do broadcasting during operations instead
2. If we dont do it, whenever we are doing classifier free guidance, we will have to manually double the size of the tensors. Like this, we dont have to worry about it, as it will just correctly broadcast.

File: torchtitan/experiments/flux/sampling.py (hunk 2 above)
Comment (CarlosGomes98) [reply torchtitan/experiments/flux/sampling.py:234]: My logic is as follows:

Previously, since we were dealing with just 1 input, we didnt have to do this, as pred would end up with a bsz of 1. In the case of classifier_free_guidance, `latents` would have a bsz of 2. Since pred has a bsz of 1, torch would broadcast this and it would work.

However, now, we can support batch sizes > 1, so pred will end up with a batch size = 1/2 of the bsz of latents, which in general will not be 1. Thus, we cannot benefit from broadcasting anymore, and have to do this repeat manually ourselves.

File: torchtitan/experiments/flux/run_inference.sh (hunk 11 above)
Comment (CarlosGomes98) [reply torchtitan/experiments/flux/run_inference.sh:15]: I would argue making it explicitly required makes using the script clearer and less error prone, at the cost of some user friendliness. But I understand that point as well. I'll make the change

File: torchtitan/experiments/flux/scripts/infer.py (hunk 12 above)
Comment (CarlosGomes98) [reply torchtitan/experiments/flux/scripts/infer.py:145]: Thats true. The tricky part is placing all the tensors back in the same order so we can match them up with the prompts. Because of the padding involved its not super straight forward, but I'm sure there's a way to do it while having each rank write its own images. Just might take some more thought

File: torchtitan/experiments/flux/scripts/infer.py (hunk 13 above)
Comment (CarlosGomes98) [reply torchtitan/experiments/flux/scripts/infer.py:31]: The only difference here is handling the batching. This could be handled by a method in the trainer, but I previously had it that way and refactored it out after discussions in https://github.com/pytorch/torchtitan/pull/1205 

I think both are valid, and its a matter of a design decision for torchtitan

File: torchtitan/experiments/flux/scripts/infer.py (hunk 12 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/scripts/infer.py:145]: An easier way to bypass padding is to require the prompts file having length divisible by DP degree, or world size. Users of this script can manually add empty rows if needed.

File: torchtitan/experiments/flux/scripts/infer.py (hunk 13 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/scripts/infer.py:31]: I see. Sounds OK to me.

File: torchtitan/experiments/flux/sampling.py (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/experiments/flux/sampling.py:208]: Ah the reasonning makes sense to me.

It feels to me that, if the result of `latent_pos_enc` is always identical for all samples in a batch, we probably should just remove the `bsz` as input arg and not worry about batch at all until its broadcast, instead of hardcoding `bsz=1` at multiple places.
https://github.com/pytorch/torchtitan/blob/ed2bbc07dda35ce26187bb0d743115381e884b35/torchtitan/experiments/flux/utils.py#L123

**Comment by wwwjn** [comment]
@CarlosGomes98 Do you want to continue working on this PR a little bit? I tried the inference script on dev machine, it works very well
